packages/node_modules/@webex/plugin-meetings/src/peer-connection-manager/index.js

```diff
@@ -492,4 +447,104 @@ pc.close = (peerConnection) => {
     });
 };
 
+
+pc.setPeerConnectionEvents = (meeting) => {
+  // In case ICE fail
+  const {peerConnection} = meeting.mediaProperties;
+
+  const connectionFailed = () => {
+    if (meeting.reconnectionManager.iceState.resolve) {
+      // DISCONNECTED state triggers first then it goes to FAILED STATE
+      // sometimes the failed state can happen before 10 seconds (Which is the timer for the reconnect for ice disconnect)
+      meeting.reconnectionManager.iceState.resolve();
+    }
+
+    meeting.reconnect({networkDisconnect: true});
+    Metrics.postEvent({
+      event: eventType.ICE_END,
+      meeting,
+      data: {
+        canProceed: false,
+        errors: [
+          Metrics.generateErrorPayload(
+            2004, false, error.name.MEDIA_ENGINE
+          )]
+      }
+    });
+  };
+
+  peerConnection.oniceconnectionstatechange = () => {
+    LoggerProxy.logger.info('Meeting:util#setPeerConnectionEvents --> ICE STATE CHANGE.');
+    switch (peerConnection.iceConnectionState) {
+      case ICE_STATE.CHECKING:
+        LoggerProxy.logger.info('Meeting:util#setPeerConnectionEvents --> ICE STATE CHECKING.');
+        Metrics.postEvent({event: eventType.ICE_START, meeting});
+        break;
+      case ICE_STATE.COMPLETED:
+        LoggerProxy.logger.info('Meeting:util#setPeerConnectionEvents --> ICE STATE COMPLETED.');
+        break;
+      case ICE_STATE.CONNECTED:
+        // Ice connection state goes to connected when both client and server sends STUN packets and
+        // Established connected between them. Firefox does not trigger COMPLETED and only trigger CONNECTED
+        Metrics.postEvent({event: eventType.ICE_END, meeting});
+        meeting.setNetworkStatus(NETWORK_STATUS.CONNECTED);
+        meeting.reconnectionManager.iceReconnected();
+        LoggerProxy.logger.info('Meeting:util#setPeerConnectionEvents --> ICE STATE CONNECTED.');
+        break;
+      case ICE_STATE.CLOSED:
+        LoggerProxy.logger.info('Meeting:util#setPeerConnectionEvents --> ICE STATE CLOSED.');
+        break;
+      case ICE_STATE.DISCONNECTED:
+        meeting.setNetworkStatus(NETWORK_STATUS.DISCONNECTED);
+        meeting.reconnectionManager.waitForIceReconnect()
+          .catch(() => {
+            LoggerProxy.logger.info('Meeting:util#setPeerConnectionEvents --> ICE STATE DISCONNECTED. Automatic Reconnection Timed Out.');
+
+            meeting.reconnect({networkDisconnect: true});
```

Comment:
Should we just call `connectionFailed()` here as well as assume the ice is failed since it stayed in disconnected state long enough?